contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,850 @@
+
+## API Object
+
+The following modifications will be made to the StatefulSetStatus API object.
+
+```go
+ type StatefulSetStatus struct {
+    // ObservedGeneration and Replicas fields are ommitted for brevity.
+
+ 	// CurrentTemplateRevision, if not nil, is the revision of the PodTemplate
+ 	// that was used to create Pods with ordinals in the sequence
+    // [0,CurrentRevisionReplicas).
+ 	TemplateRevision  *int64 `json:"templateRevision,omitempty"`
+ 	
+ 	// TargetTemplateRevision, if not nil, is the revision of the PodTemplate 
+ 	// that was used to create Pods with ordinals in the sequence 
+    // [Replicas - TargetRevisionReplicas, Replicas).
+ 	TargetTemplateRevision *int64 `json:"targetTemplateRevision,omitempty"`
+ 	
+ 	// ReadyReplicas is the current number of Pods, created by the StatefulSet
+    // controller, that have a Status for Running and a Ready Condition.
+ 	ReadyReplicas int32 `json:"readyReplicas,omitempty"`
+ 	
+ 	// CurrentRevisionReplicas is the number of Pods created by the StatefulSet 
+    // controller from the PodTemplateSpec indicated by CurrentTemplateRevision.
+ 	CurrentReplicas int32 `json:"currentReplicas,omitempty"`
+ 	
+ 	// TargetRevisionReplicas is the number of Pods created by the StatefulSet
+    // controller from the PodTemplateSpec indicated by TargetTemplateRevision.
+ 	TargeReplicas int32 `json:"taretReplicas,omitempty"`
+}
+```
+
+The following modifications will be made to the StatefulSetSpec API object.
+
+```go
+type StatefulSetSpec struct {
+    // Replicas, Selector, Template, VolumeClaimsTemplate, and ServiceName 
+    // ommitted for brevity.
+	v1.PodTemplateSpec `json:"template"`
+
+	// TemplateRevision is a monotonically increasing, 64 bit, integer used to 
+	// indicate the version of the of the PodTemplateSpec. If nil, the 
+	// StatefulSetController has not initialized its revision history,
+	// change tracking is not enabled, and all Pods will be created from 
+	// Template.
+	TemplateRevision  *int64 `json:"templateRevision"`
+
+	// RevisionParition paritions the Pods in the StatefulSet by ordinal such 
+    // that all Pods with a lower ordinal will be created from the PodTemplate that
+    // represents the current revision of the StatefulSet's revision history and 
+    // all Pods with an a greater or equal ordinal will be created from the 
+    // PodTemplate that represents the target revision of the StatefulSet's 
+    // revision history.
+	RevisionPartition *int32 `json:"revisionPartition,omitempty`
+
+	// RevisionHistoryDepth is the maximum number of PodTemplates that will 
+	// be maintained in the StatefulSet's revision history. It must be at 
+	// least two.
+	RevisionHisotryDepth int32 `json:historyRevisionDepth,omitempty`
+}
+```
+
+Additionally, we introduce the following constants.
+
+```go 
+// StatefulSetPodTemplateLabel is the label applied to a PodTemplate to allow
+// the StatefulSet controller to select the PodTemplates in its revision 
```

Comment:
To select them. OwnerReference is needed as well to avoid revision history overlap.